### ISSUE TITLE
Deploying EHRBase with Postgres Database without Extensions

### ISSUE THREAD

**Issue by sidharthramesh** [ehrbase/ehrbase#508]
## Configuration information

- EHRbase version: v0.16.5
- PostgreSQL version: 13.3
- Operating System version: Linux

## Steps to reproduce ## 

I'm trying to get EHRBase running with a Postgres container without any extensions. This is the most common scenario while deploying to the cloud with managed databases. I've tried to follow the guide by @chevalleyc [here](https://github.com/ehrbase/ehrbase/blob/262cfbb354d8adca75c5cd8ab7f9cb9fbf98adc7/base/db-setup/README.md).

This is the docker-compose.yaml:
```
version: "3"
services:
  ehrbase:
    image: ehrbase/ehrbase:next
    ports:
      - 8080:8080
    environment:
      SPRING_DATASOURCE_URL: "jdbc:postgresql://db:5432/ehrbase"
      SPRING_DATASOURCE_USERNAME: admin
      SPRING_DATASOURCE_PASSWORD: admin
      SERVER_AQLCONFIG_USEJSQUERY: "false"
    depends_on:
      - db
  db:
    image: postgres
    volumes:
      - db-data:/var/lib/postgresql/data
    environment:
      POSTGRES_PASSWORD: admin
      POSTGRES_USER: admin
      POSTGRES_DB: ehrbase
volumes:
  db-data:
```
There was some discrepancy in the name of the variable to disable using the JsQuery extension. According to [the guide](https://github.com/ehrbase/ehrbase/blob/262cfbb354d8adca75c5cd8ab7f9cb9fbf98adc7/base/db-setup/README.md), it's `server.aql.use-jsquery`, but in the default [application.yaml](https://github.com/ehrbase/ehrbase/blob/develop/application/src/main/resources/application.yml) it's `server.aqlConfig.useJsQuery` which translates to the environment variable `SERVER_AQLCONFIG_USEJSQUERY` in the above docker-compose.yml

## Actual result ##

Running migrations on the postgres database fails because postgres does not understand what "ext" means:
```
ehrbase_1     | 2021-05-23 13:56:18.376  INFO 8 --- [           main] o.f.core.internal.command.DbMigrate      : Current version of schema "ehr": 1
ehrbase_1     | 2021-05-23 13:56:18.516  INFO 8 --- [           main] o.f.core.internal.command.DbMigrate      : Migrating schema "ehr" to version 2 - ehr
db_1          | 2021-05-23 13:56:18.535 UTC [32] ERROR:  schema "ext" does not exist at character 160
db_1          | 2021-05-23 13:56:18.535 UTC [32] STATEMENT:  -- storeComposition common_im entities
db_1          |         -- CREATE TABLE "system" ---------------------------------------
db_1          |         CREATE TABLE ehr.system (
db_1          |                 id UUid PRIMARY KEY DEFAULT ext.uuid_generate_v4(),
db_1          |                 description TEXT NOT NULL,
db_1          |                 settings TEXT NOT NULL
db_1          |          )
ehrbase_1     | 2021-05-23 13:56:18.586 ERROR 8 --- [           main] o.f.core.internal.command.DbMigrate      : Migration of schema "ehr" to version 2 - ehr failed! Changes successfully rolled back.
ehrbase_1     | 2021-05-23 13:56:18.602  WARN 8 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 
'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Invocation of init method failed; nested exception is org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException:
ehrbase_1     | Migration V2__ehr.sql failed
ehrbase_1     | ----------------------------
ehrbase_1     | SQL State  : 3F000
ehrbase_1     | Error Code : 0
ehrbase_1     | Message    : ERROR: schema "ext" does not exist
ehrbase_1     |   Position: 160
ehrbase_1     | Location   : db/migration/V2__ehr.sql (/file:/ehrbase.jar!/BOOT-INF/lib/base-0.16.5.jar!/db/migration/V2__ehr.sql)
ehrbase_1     | Line       : 44
ehrbase_1     | Statement  : -- storeComposition common_im entities
ehrbase_1     | -- CREATE TABLE "system" ---------------------------------------
ehrbase_1     | CREATE TABLE ehr.system (
ehrbase_1     |         id UUid PRIMARY KEY DEFAULT ext.uuid_generate_v4(),
ehrbase_1     |         description TEXT NOT NULL,
ehrbase_1     |         settings TEXT NOT NULL
ehrbase_1     |  )
```
All of the subsequent migrations also fail due to the same reason. I've only included these logs here for brevity. I've included the entire logs in [this pastebin](https://pastebin.com/xmshZC1G) if you're interested.

## Expected result (Acceptance Criteria) ##

The migrations must be successful on the base Postgres image without any extensions. All existing functionality of EHRBase should be present.

## Definition of Done ##

- [ ] The defect is checked by an unit or an integration test (Robot)
- [ ] Merge Request approved
- [ ] Unit tests passed
- [ ] Build without errors
- [ ] Release notes prepared
- [ ] No additional runtime warnings

**Comment by sidharthramesh** [comment]
I have executed the following SQL statements as in the [prepare-database.sh script](https://github.com/ehrbase/docker/blob/master/dockerfiles/scripts/prepare-databases.sh):
```
CREATE SCHEMA IF NOT EXISTS ehr AUTHORIZATION admin;  
CREATE SCHEMA IF NOT EXISTS ext AUTHORIZATION admin;  
CREATE EXTENSION IF NOT EXISTS "uuid-ossp" SCHEMA ext;  
CREATE EXTENSION IF NOT EXISTS "ltree" SCHEMA ext;  
ALTER DATABASE ehrbase SET search_path to "admin",public,ext;  
GRANT ALL ON ALL FUNCTIONS IN SCHEMA ext TO admin;
```

I have no extension installed, so have NOT executed these statements:
```
CREATE EXTENSION IF NOT EXISTS "temporal_tables" SCHEMA ext;  
CREATE EXTENSION IF NOT EXISTS "jsquery" SCHEMA ext;  
```

And EHRBase fails to startup due to the missing `versioning()` function.
```
db_1       | 2021-05-24 08:45:00.030 UTC [32] ERROR:  function ext.versioning() does not exist
db_1       | 2021-05-24 08:45:00.030 UTC [32] STATEMENT:  CREATE TRIGGER versioning_trigger BEFORE INSERT OR UPDATE OR DELETE ON ehr.status  
db_1       |    FOR EACH ROW EXECUTE PROCEDURE ext.versioning('sys_period', 'ehr.status_history', true)
ehrbase_1  | 2021-05-24 08:45:00.051 ERROR 9 --- [           main] o.f.core.internal.command.DbMigrate      : Migration of schema "ehr" to version 2 - ehr failed! Changes successfully rolled back.
ehrbase_1  | 2021-05-24 08:45:00.057  WARN 9 --- [           main] ConfigServletWebServerApplicationContext : Exception encountered during context initialization - cancelling refresh attempt: org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'flywayInitializer' defined in class path resource [org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration$FlywayConfiguration.class]: Invocation of init method failed; nested exception is org.flywaydb.core.internal.command.DbMigrate$FlywayMigrateException:
ehrbase_1  | Migration V2__ehr.sql failed
ehrbase_1  | ----------------------------
ehrbase_1  | SQL State  : 42883
ehrbase_1  | Error Code : 0
ehrbase_1  | Message    : ERROR: function ext.versioning() does not exist
ehrbase_1  | Location   : db/migration/V2__ehr.sql (/file:/ehrbase.jar!/BOOT-INF/lib/base-0.16.5.jar!/db/migration/V2__ehr.sql)
ehrbase_1  | Line       : 146
ehrbase_1  | Statement  : CREATE TRIGGER versioning_trigger BEFORE INSERT OR UPDATE OR DELETE ON ehr.status
ehrbase_1  | FOR EACH ROW EXECUTE PROCEDURE ext.versioning('sys_period', 'ehr.status_history', true)
ehrbase_1  |
ehrbase_1  | 2021-05-24 08:45:00.060  INFO 9 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown initiated...
ehrbase_1  | 2021-05-24 08:45:00.102  INFO 9 --- [           main] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Shutdown completed.
ehrbase_1  | 2021-05-24 08:45:00.116  INFO 9 --- [           main] o.apache.catalina.core.StandardService   : Stopping service [Tomcat]
ehrbase_1  | 2021-05-24 08:45:00.152  INFO 9 --- [           main] ConditionEvaluationReportLoggingListener : 
ehrbase_1  |
ehrbase_1  | Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
ehrbase_1  | 2021-05-24 08:45:00.164 ERROR 9 --- [           main] o.s.boot.SpringApplication               : Application run failed
```
How do I get this `ext.versioning()` function without installing the `temporal_tables` extension?
I'm using Postgresql v13.3, with docker tag `postgres:13.3`.  Should I run something [like this](https://github.com/jasonk/postgresql-versioning/blob/master/postgresql-versioning.sql)? (The repository seems like a hobby project) 

I'm trying to deploy this on Digital Ocean, and only [some extensions](https://docs.digitalocean.com/products/databases/postgresql/resources/supported-extensions/) are supported. The same goes for [AWS RDS](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html). `uuid-ossp` and `ltree` are supported widely.

Any idea on how to get the `versioning()` function without `temporal_tables`?

**Comment by jakesmolka** [comment]
Hey @sidharthramesh! Have you read my blog post on [how to deploy EHRbase on Azure](https://blog.surrounded.space/post/2021/02/22/deploying-ehrbase-on-azure/)? It was a bit bumpy, but I manged to run the DB as a managed instance, without the extensions.
I'll have a thorough look on your report here later.

**Comment by sidharthramesh** [comment]
I've run the scripts, but seems like some of these migration scripts have the "ehrbase" user/role hard-coded.

https://github.com/ehrbase/ehrbase/blob/262cfbb354d8adca75c5cd8ab7f9cb9fbf98adc7/base/src/main/resources/db/migration/V16__fix_plpgsql_js_functions_composition_uid.sql#L17

Should I open a new issue?

**Comment by sidharthramesh** [comment]
I have applied the migrations from the SQL script and, the server has started up without any issues now. So yay! 

I tried to manually apply the [db-triggers-script](https://github.com/ehrbase/ehrbase/blob/develop/base/db-setup/cloud-db-triggers-setup.sql), but seems like EHRBase takes care of this in the migration process (all the triggers already existed).

Considering that the extensions are no longer required, wouldn't it be better to integrate all of these into the EHRBase migration scripts and make this the default? The `ehrbase` docker container can run without depending on the `ehrbase-db` container. This will also make EHRBase cloud-ready by default. 

What do you think?

**Comment by sidharthramesh** [comment]
Creating an EHR by POST to `http://localhost:8080/ehrbase/rest/openehr/v1/ehr` causes these error logs in the Postgres db:
```
db_1       | 2021-05-25 17:38:09.344 UTC [32] ERROR:  column "sys_period" of relation "audit_details" does not exist
db_1       | 2021-05-25 17:38:09.344 UTC [32] CONTEXT:  PL/pgSQL function versioning() line 39 at RAISE
db_1       | 2021-05-25 17:38:09.344 UTC [32] STATEMENT:  insert into "ehr"."audit_details" ("system_id", "committer", "time_committed", "time_committed_tzid", "change_type", "description") values (cast($1 as uuid), cast($2 as uuid), cast($3 as timestamp), $4, $5::"ehr"."contribution_change_type", $6) returning "ehr"."audit_details"."id"
```
EHRBase response:
```
{
  "error": "Internal Server Error: Could not create an EHR with given parameters.",
  "status": "Internal Server Error"
}
```
No other logs from EHRBase.

**Comment by jakesmolka** [comment]
Great it worked!

I totally agree. We also have an internal ticket for the migration to the postgres native versioning, instead of the temporal table extension. But it is not a top priority currently. I'll talk with the team about it.

**Comment by birgerhaarbrandt** [comment]
@jakesmolka @sidharthramesh Would be appropriate to close this ticket?

**Comment by sidharthramesh** [comment]
@birgerhaarbrandt, I think EHRbase should automatically apply migrations that have not been applied on startup. So, it'll be better to close this after that feature has been added. If you have an internal ticket for that, then no worries!

**Comment by wlad** [comment]
This can be closed, since ehrbase/ehrbase-postgres:13.3 no postgres extensions required any more, 

use this [script](https://github.com/ehrbase/docker/blob/master/dockerfiles/scripts/db-setup.sql) or [this one](https://github.com/ehrbase/ehrbase/blob/develop/base/db-setup/cloud-db-setup.sql) if you use a 'naked' (non ehrbse) postgres image
